src/core/gui/inputdevices/TouchInputHandler.cpp

```diff
@@ -22,64 +22,60 @@ TouchInputHandler::TouchInputHandler(InputContext* inputContext): AbstractInputH
 auto TouchInputHandler::handleImpl(InputEvent const& event) -> bool {
     bool zoomGesturesEnabled = inputContext->getSettings()->isZoomGesturesEnabled();
 
-    // Don't handle more then 2 inputs
-    if (this->primarySequence && this->primarySequence != event.sequence && this->secondarySequence &&
-        this->secondarySequence != event.sequence) {
-        return false;
-    }
-
     if (event.type == BUTTON_PRESS_EVENT) {
-        // Start scrolling when a sequence starts and we currently have none other
-        if (this->primarySequence == nullptr && this->secondarySequence == nullptr) {
-            this->primarySequence = event.sequence;
-
-            // Set sequence data
-            sequenceStart(event);
-        }
-        // Start zooming as soon as we have two sequences.
-        else if (this->primarySequence && this->primarySequence != event.sequence &&
-                 this->secondarySequence == nullptr) {
-            this->secondarySequence = event.sequence;
-
-            // Set sequence data
-            sequenceStart(event);
-
-            // Even if zoom gestures are disabled,
-            // this is still the start of a sequence. Just
-            // don't start zooming.
-            if (zoomGesturesEnabled) {
-                zoomStart();
+        if (invalidActive.empty() && event.sequence != primarySequence && event.sequence != secondarySequence) {
+            // All touches are previously valid and we did not miss an end/cancel event for the current touch
+            if (primarySequence && secondarySequence) {
+                // All touches become invalid
+                invalidActive.insert({primarySequence, secondarySequence, event.sequence});
+                primarySequence = secondarySequence = nullptr;
+            } else {
+                if (primarySequence) {
+                    secondarySequence = event.sequence;
+                } else {
+                    primarySequence = event.sequence;
+                }
+                sequenceStart(event);
+                if (secondarySequence) {
+                    zoomStart();
+                }
             }
+        } else {
+            invalidActive.insert(event.sequence);
         }
+        return true;
     }
 
-    if (event.type == MOTION_EVENT && this->primarySequence) {
-        if (this->primarySequence && this->secondarySequence && zoomGesturesEnabled) {
-            zoomMotion(event);
-        } else if (event.sequence == this->primarySequence) {
+    if (event.type == MOTION_EVENT) {
+        if (primarySequence == event.sequence && !secondarySequence) {
             scrollMotion(event);
-        } else if (this->primarySequence && this->secondarySequence) {
-            sequenceStart(event);
+            return true;
+        } else if (event.sequence && (primarySequence == event.sequence || secondarySequence == event.sequence)) {
```

Comment:
Maybe adding an assertion check here might be good, just in case...
```suggestion
        } else if (event.sequence && (primarySequence == event.sequence || secondarySequence == event.sequence)) {
        xoj_assert(primarySequence);
```

Also, if event.sequence could really be 0, then we need to check it's not in the first test above as well.